### ISSUE TITLE
page cache & django.contrib.messages

### ISSUE THREAD

**Issue by ollb** [django-cms/django-cms#5725]
The page cache seems not compatible with `django.contrib.messages`. Pages are cached with messages if any and displayed without taking current available messages in account.

As a workaround, we disabled page cache in settings (but we loose cache benefit in all other legitimate cases):

```
CMS_PAGE_CACHE = False
```

Fixes:
- A specific fix could be to check messages in set_page_cache/get_page_cache functions.
- A more flexible fix could be to allow to write a custom `_cache_page_key` function configurable in settings (and also prevent cache use in case of `None` key). And it would be wise to advise developer to only add a suffix in those custom implementations (or return `None`).

Both could be done. What would you say?

**Comment by yakky** [comment]
Thanks for reporting this, we're working to improve the granularity of the cache control on django CMS.
The latter approach seems more sensible, as it allows to make use of the page cache across very different situation.
Are you willing to contribute this feature?

**Comment by czpython** [comment]
@yakky I don't consider this to be a feature, seems to me that if we support page level caching then we should account for builtin django features like the messages framework. The only feature I wouldn't support out of the box (yet) is csrf handling with caching as this is much more complex and can vary depending on the project.

Instead of building the generic solution now, I advice we first address the problem on 3.4.x and then we can coordinate a more generic solution to land on 3.5.x. Honestly I'm not fond of the idea to extend the existing page level caching and add more settings/granularity, we should first check if there's really a strong need for this.

**Comment by ollb** [comment]
Yes, I think it should be considered as a bug in 3.4 as messages framework it a very common feature in django.

Providing page key function customization is clearly an enhancement. But, it is very annoying to globally disable page cache just because whatever corner case has not been handled in django-cms. We create another issue ?

@yakky I can provide pull request on both issues (we are currently disabling page cache by default just because of this...).

**Comment by czpython** [comment]
@ollb 
If you're able to, please provide a fix for the messages issue so we can land it on 3.4.x
Regarding the page key function customization, I suggest to create a post in our [developers](https://groups.google.com/forum/#!forum/django-cms-developers) group so we can discuss if this is really needed.

**Comment by ollb** [comment]
@czpython I've created a pull request to fix compatibility between page cache and django.contrib.messages (integration tests seem buggy: they are complaining about coverage nonexistence "coverage: command not found").
